### DISCUSSION TITLE
VB.net options

### DISCUSSION THREAD

**Discussion by secureitall** [webspiderteam/Bluetooth-DMM-For-Windows#39] [General]
Hi all im looking to write a small app to show the values from the meter however i dont know c# and looking through the code and using telerik i cant get anything useable.
All im looking to do is connect to the meter, see what mode its in, ie volts, mA etc and get the reading.

All the examples for BT LE dont seem to help me connect to the meter but the complied application works wonderfully.

Many Thanks

J

**Comment by webspiderteam** [comment]
hi, latest version can be complicated but if you want to do it yourself you can use https://github.com/webspiderteam/Bluetooth-DMM-For-Windows/releases/tag/WPFAlpha_1.04 this version. this version has only displaying codes and connection codes. ofcourse there is issues with it but I can help you if you want. BTW whick device do you have?

**Comment by secureitall** [comment]
Hi Many Thanks

 

Ive downloaded the alpha and will work through it. I only want to connect to one device and get the value etc.

 

The device I have is an AN9002 got it because of your code assuming I could connect to it lol

 

J

**Comment by webspiderteam** [comment]
Ok. I am working on it. Probably I will send  a useable source code for you tomorrow.

**Comment by secureitall** [comment]
Oh Wow

Ive been looking at converting some of it but ive found some things appear to not be possible in vb.net and I don’t have the ability

 

J

**Comment by webspiderteam** [comment]
Can you explain me ?
where will you use it?
how its should look like ?
why do you need smaller app?
and last one is can you show me a picture of how it should look?

**Comment by secureitall** [comment]
Hi

Its to get a value from the meter to show if it’s a good or a bad value.

Ie battery results if above 1.3v then green and pass below read and fail.

I don’t need all the discovery parts or the pretty display, just a way to connect  to the device by address in vb.net and a means to display the actual reading on the meter (and its current selected range) in a textbox..

 

J

**Comment by webspiderteam** [comment]
but is already has most of them
![image](https://github.com/webspiderteam/Bluetooth-DMM-For-Windows/assets/85828505/0d6e7694-ecd9-4204-8257-968e25550ad7)

when you set necessary settings it will connect without asking anything or or you dont need to connecttion things

**Comment by secureitall** [comment]
Agreed, your pp works perfectly but I need to get the values from the meter into my app so I can manipulate the readings obtained.

All of the examples I can find that can connect to say a heart rate monitor uses DeviceInformation.FindAllAsync(BluetoothDevice.GetDeviceSelector())

And this doesn’t list the an9002 meter.

 

Ive looked at the Bluetooth le explorer and the meter shows up in that and displays values (in hex)

Plus your app fully connects to the meter and does everything.

**Reply by webspiderteam** [reply]
it has Mqtt ability too. Did you see it? it gets values from meter and its very simple app to make anything as you want. you can find it here https://github.com/webspiderteam/Bluetooth_DMM_For_Windows_MQTT_Test_Project

also for bluetooth connection ability I have used that sample code 
[HeartRateLE-master (1).zip](https://github.com/webspiderteam/Bluetooth-DMM-For-Windows/files/11848521/HeartRateLE-master.1.zip)
this was my starting point for connecting to meter. You can inspect it for understanding code.

**Comment by secureitall** [comment]
Yes I did consider setting up a mqtt server and listening to that but thought it was a little overcomplicated for what im trying to do. I will look at this though.

 

I have looked at the heartrate monitor sample you suggested and used Telerik to convert sections from c# to vb.net

 

However

The first bit I have tried to translate and work though gives me a load of errors

 

 

Public Async Function ConnectAsync(ByVal deviceId As String) As Task(Of ConnectionResult)

        _heartRateDevice = Await BluetoothLEDevice.FromIdAsync(deviceId)

 

        If _heartRateDevice Is Nothing Then

            Return New Schema.ConnectionResult() With {

                .IsConnected = False,

                .ErrorMessage = "Could not find specified heart rate device"

            }

        End If

 

        If Not _heartRateDevice.DeviceInformation.Pairing.IsPaired Then

            _heartRateDevice = Nothing

            Return New Schema.ConnectionResult() With {

                .IsConnected = False,

                .ErrorMessage = "Heart rate device is not paired"

            }

        End If

 

        _heartRateDevice.ConnectionStatusChanged -= DeviceConnectionStatusChanged

        _heartRateDevice.ConnectionStatusChanged += DeviceConnectionStatusChanged

        Dim isReachable = Await GetDeviceServicesAsync()

 

        If Not isReachable Then

            _heartRateDevice = Nothing

            Return New Schema.ConnectionResult() With {

                .IsConnected = False,

                .ErrorMessage = "Heart rate device is unreachable (i.e. out of range or shutoff)"

            }

        End If

 

        Dim characteristicResult As CharacteristicResult

        characteristicResult = Await SetupHeartRateCharacteristic()

        If Not characteristicResult.IsSuccess Then Return New Schema.ConnectionResult() With {

            .IsConnected = False,

            .ErrorMessage = characteristicResult.Message

        }

        DeviceConnectionStatusChanged(_heartRateDevice, Nothing)

        Return New Schema.ConnectionResult() With {

            .IsConnected = _heartRateDevice.ConnectionStatus = BluetoothConnectionStatus.Connected,

            .Name = _heartRateDevice.Name

        }

 

I was hoping that I could see the code to connect to the meter from this then work on getting the value from it.

Understand that once I have got the values I have to convert them into something readable

 

Thanks for your help so far

 

James

**Comment by secureitall** [comment]
I think the first step I need to achieve is to connect to the meter by its address and get the hex values it streams. Then I can look at how to convert that using the information on Home · ludwich66/Bluetooth-DMM Wiki (github.com) <https://github.com/ludwich66/Bluetooth-DMM/wiki> 

 

J

**Comment by webspiderteam** [comment]
there was a vb winform example. did you see it?

**Reply by secureitall** [reply]
Ah, no i didnt I was looking at the c# code and trying to decipher

**Reply by webspiderteam** [reply]
https://www.vb-paradise.de/index.php/Thread/132559-Bluetooth-Ansprechen-Ansteuern-von-BLE-Ger%C3%A4ten/ here is an example. its not english but you can use translate

**Comment by webspiderteam** [comment]
btw you dont need to convert dll file you can use it as it is. you ony need to make main window

**Reply by secureitall** [reply]
Sorry I dont understand?

**Reply by webspiderteam** [reply]
you can use my bluetooth.dmm.dll as referance and you can use it in vb.net

**Reply by webspiderteam** [reply]
[WpfApp3.zip](https://github.com/webspiderteam/Bluetooth-DMM-For-Windows/files/11850939/WpfApp3.zip)
you can use this as start point with dll. this dll also decodes hex data so you dont need to try decode just use.

**Comment by secureitall** [comment]
That looks to be an amazing option. If I add the dll as a reference is there any docs on how to use it from vb?

**Reply by webspiderteam** [reply]
sorry there is no docs but codes should be like that. but not sure is it true or not  

```
Imports System.ComponentModel
Imports HeartRateLE.Bluetooth
Imports HeartRateLE.Bluetooth.Events
Imports HeartRateLE.Bluetooth.Schema
Imports Windows.Devices.Bluetooth
Imports Windows.Devices.Enumeration

Class MainWindow
    Private _heartRateMonitor As HeartRateLE.Bluetooth.HeartRateMonitor
    Private Property SelectedDeviceId As String
    Private Property SelectedDeviceName As String


    Private Sub HrParserOnValueChanged(sender As Object, e As RateChangedEventArgs)
        txtData.Content = e.MyGattCData
        txtSymbol.Content = e.MyGattCDataSymbol
    End Sub

    Private Sub HrDeviceOnDeviceConnectionStatusChanged(sender As Object, e As ConnectionStatusChangedEventArgs)
        MessageBox.Show("gg")

    End Sub

    Private Async Function Button_ClickAsync(sender As Object, e As RoutedEventArgs) As Task Handles button.Click

        Debug.WriteLine("Starting enumeration...")
        Dim dp = New DevicePicker()
        Dim rec0 = New Windows.Foundation.Rect(10, 10, 400, 500)
        dp.Filter.SupportedDeviceSelectors.Add(BluetoothLEDevice.GetDeviceSelectorFromPairingState(False))
        'dp.Filter.SupportedDeviceSelectors.Add(BluetoothLEDevice.GetDeviceSelectorFromDeviceName("Bluetooth DMM"))
        'dp.Show(rec0)

        Dim mydev As DeviceInformation = Await dp.PickSingleDeviceAsync(rec0, Windows.UI.Popups.Placement.Above)
        If mydev IsNot Nothing Then
            Await _heartRateMonitor.ConnectAsync(mydev.Id).ConfigureAwait(False)
            'Await mydev.Pairing.PairAsync(DevicePairingProtectionLevel.EncryptionAndAuthentication)
            Dim connectResult = Await _heartRateMonitor.ConnectAsync(mydev.Id)
            MessageBox.Show("gg")
            If (Not connectResult.IsConnected) Then
                MessageBox.Show(connectResult.ErrorMessage)
            End If
        End If


    End Function

    Private Sub MainWindow_Loaded(sender As Object, e As RoutedEventArgs) Handles Me.Loaded
        _heartRateMonitor = New HeartRateMonitor()

        ' we should always monitor the connection status
        RemoveHandler _heartRateMonitor.ConnectionStatusChanged, AddressOf HrDeviceOnDeviceConnectionStatusChanged
        AddHandler _heartRateMonitor.ConnectionStatusChanged, AddressOf HrDeviceOnDeviceConnectionStatusChanged


        ' we can create value parser And listen for parsed values of given characteristic
        'HrParser.ConnectWithCharacteristic(HrDevice.HeartRate.HeartRateMeasurement)
        RemoveHandler _heartRateMonitor.RateChanged, AddressOf HrParserOnValueChanged
        AddHandler _heartRateMonitor.RateChanged, AddressOf HrParserOnValueChanged
        'connectResult = Await _heartRateMonitor.ConnectAsync("")
    End Sub
End Class
```

**Comment by secureitall** [comment]
That’s wonderful

Where do I find the dll’s to reference for the import statements?

Ive looked in the bin folder of the previous source and there arnt there and I cant build it as there are errors in visual studio

 

J

**Comment by webspiderteam** [comment]
@secureitall  Here is project that I made in vb.net. download and make your changes as you wish. 

[BluetoothDMM_VB.NET.zip](https://github.com/webspiderteam/Bluetooth-DMM-For-Windows/files/11860004/BluetoothDMM_VB.NET.zip)

its very basic stuffs. I only connected and got the values. also deceded them too...

**Comment by secureitall** [comment]
Wow

Wonderful this seems to make more sense to me.

Ill let you know how I get on

 

Superb Many Thanks

 

James

**Reply by webspiderteam** [reply]
that would be great :)
